### DISCUSSION TITLE
Support OpenTelemetry-based instrumentation (openllmetry and others)

### DISCUSSION THREAD

**Discussion by gyliu513** [langfuse/langfuse#2509] [Ideas]
### Describe the feature or potential improvement

openllmetry can provide many instrumentation code for vectordb, LLMs, LLM orchestration platforms etc, and traceloop sdk is kind of otel collector and can expose metrics, logs(did not implement yet), tracing to a third party observability platform. Langfuse is a good candidate to act as a 3rd party LLM observability platform.

> [!NOTE]  
> Documentation on Otel Support in Langfuse is available here: https://langfuse.com/integrations/native/opentelemetry

**Comment by marcklingen** [comment]
Thanks for sharing! I think generally supporting the forming opentelemetry conventions (genai/llm) is super interesting as they mature. There are currently a number of instrumentation implementations with slightly different conventions as far as I understand the notes of the working group(s). I'd love to understand your perspective on this!

I think adding an OTel collector to the langfuse platform seems the most promising way to support OTel-based instrumentation as long as the formats/semantics across instrumentation libraries are starting to be standardized/stable enough.

**Comment by gyliu513** [comment]
> I think adding an OTel collector to the langfuse platform seems the most promising way to support OTel-based instrumentation as long as the formats/semantics across instrumentation libraries are starting to be standardized/stable enough.

@marcklingen Yes, OTel collector will be the best option for this. I was wondering maybe even before the formats/semantics across instrumentation libraries become mature, you can probably still able to build some OTel collectors to integrate with langfuse, as the formats/semantics across instrumentation libraries did not impact much for the implementation. Embracing OTEL can also help enlarge the ecosystem for langfuse as well.

> There are currently a number of instrumentation implementations with slightly different conventions as far as I understand the notes of the working group(s). I'd love to understand your perspective on this!

Yes, but the OTEL community is trying to unify this via the LLM working group, there is already a initial draft spec for LLM semantic convention at https://github.com/open-telemetry/semantic-conventions/tree/main/docs/gen-ai

**Reply by marcklingen** [reply]
> I was wondering maybe even before the formats/semantics across instrumentation libraries become mature, you can probably still able to build some OTel collectors to integrate with langfuse, as the formats/semantics across instrumentation libraries did not impact much for the implementation

I am unsure about this one as
- we will not be able to support all Langfuse features while they are not yet covered by the conventions
- depending on the speed of the development of the conventions, once we add an initial otel collector (with basic features), there will probably be great interest to support vendor-specific implementation aspects while the standards still develop. This leads to potential maintenance overhead and breaking backwards compatibility once conventions are in place for everyone who relied on the 3rd party vendor instrumentation.

> Yes, but the OTEL community is trying to unify this via the LLM working group, there is already a initial draft spec for LLM semantic convention at https://github.com/open-telemetry/semantic-conventions/tree/main/docs/gen-ai

I'll try to join the working group to keep in the loop with these changes, very interested in this. I am especially excited about this proposal: https://github.com/open-telemetry/community/pull/2326

**Comment by gyliu513** [comment]
Thanks @marcklingen , I totally understand your concern for backward compatibility.

> we will not be able to support all Langfuse features while they are not yet covered by the conventions
> depending on the speed of the development of the conventions, once we add an initial otel collector (with basic features), there will probably be great interest to support vendor-specific implementation aspects while the standards still develop. This leads to potential maintenance overhead and breaking backwards compatibility once conventions are in place for everyone who relied on the 3rd party vendor instrumentation.

Let me share some of my understanding, usually, if a product did not reach GA milestone, the product usually in Beta or Alpha state, and customer can have a quick try and share some feedback. Due to the product is Beta or Alpha, and it will not support running in production, but just PoC. When reaching GA, the customer may need to allow some break changes. But anyway, glad to see you are interested in OTEL AI Semantic Convention, look forward to work with you in this area 👍

**Reply by marcklingen** [reply]
Fair, we might ship something like this as a quick first POC at one point to get feedback and collaborate with everyone, labeling it as beta. However, getting to GA quickly was something we focused on with all of our other integrations as it was valued by many teams using Langfuse. In this case, we need to make clear that things can still break while the conventions change.

**Reply by marcklingen** [reply]
btw, awesome to see you being interested to contribute to the new otel instrumentation libraries

**Reply by gyliu513** [reply]
> Fair, we might ship something like this as a quick first POC at one point to get feedback and collaborate with everyone

I'm excited for this, look forward to see this :)

**Comment by mjrusso** [comment]
Just adding a note here that OTel support would also enable Langfuse to support [Magentic](https://github.com/jackmpcollins/magentic) (cc @jackmpcollins) and any library that is itself instrumented with [Pydantic Logfire](https://pydantic.dev/logfire). For Magentic in particular, see details here: https://magentic.dev/logging-and-tracing/

**Reply by marcklingen** [reply]
thanks!

**Reply by baggiponte** [reply]
logfire is superb, would be awesome to get the integration (also given that pydantic is ubiquitous).

**Comment by marcklingen** [comment]
OTel support would also allow to integrate with [Firebase GenKit](https://firebase.google.com/docs/genkit), see [instrumentation docs](https://firebase.google.com/docs/genkit/plugin-authoring-telemetry) and thread: https://github.com/orgs/langfuse/discussions/3351

cc @debkanchan

**Comment by marcklingen** [comment]
Merging thread #2043 by @baggiponte
> [openllmetry](https://github.com/traceloop/openllmetry) by Traceloop allows instrumenting your LLM applications and send data to any collector such as splunk, datadog, signoz and so on. It's a standardisation effort that's becoming quite popular and successful and I think langfuse would benefit from supporting that.
> This would mean that most applications already using openllmetry/opentelemetry could seamlessly send data to langfuse (or a different product, like the APM posthog). This could tie langfuse closer to the telemetry ecosystem. It would also reduce the duplication effort of instrumenting the same app for two different collectors.

**Reply by johnson-liang-appier** [reply]
Maybe merge this issue too?
https://github.com/langfuse/langfuse/issues/199

**Reply by marcklingen** [reply]
makes sense, let's merge all threads on this here. thank you!

**Comment by marcklingen** [comment]
Merging #199

> [pamelafox](https://github.com/pamelafox) commented [on Mar 5](https://github.com/langfuse/langfuse/issues/199#issuecomment-1979289447)
> Did you all make a decision regarding OTel integration? My team at Microsoft is curious (since Azure Monitor brings in OTel traces).
> 
> @marcklingen [marcklingen](https://github.com/marcklingen) commented [on Mar 5](https://github.com/langfuse/langfuse/issues/199#issuecomment-1979301915) • 
> Currently we see otel to be promising for some of our users (just monitor/log production traffic), whereas it does not work for others (also want to quickly iterate locally / in batches / in notebooks). Thus, we cannot yet shift all of our tracing/ingestion to OTel.
> 
> The best way forward here is most likely to build an OTel compatible receiver/processor as an alternative way to ingest data into langfuse for those who want it for their production deployments.
> 
> As generally building integrations for Langfuse is very simple, we are very open to collab on this, right now, there are higher impact features to ship for most of our users especially with a focus on evaluation and testing of LLM applications.
> 
> If you have a view on this, happy to chat with you and your team!
> 
> @[pamelafox](https://github.com/pamelafox)
> pamelafox commented [on Mar 5](https://github.com/langfuse/langfuse/issues/199#issuecomment-1979314786)
> Okay, that makes sense. I don't have a strong opinion currently, I'll look out for feedback from our Microsoft customers.
> 
> @[sudivate](https://github.com/sudivate)
> sudivate commented [on Mar 8](https://github.com/langfuse/langfuse/issues/199#issuecomment-1984823117) • 
> Hi @marcklingen, it's great to know that you're thinking about integrating OTEL into Langfuse.
> 
> A new Working Group (WG) was formed recently as part of the OpenTelemetry Semantic Conventions project, with a specific focus on LLM applications and frameworks.
> 
> If you're interested, I'd like to extend an invitation for you to join our community meeting. You can share any insights you have from your experience with OTEL adoption and provide feedback on the conventions we're developing for this project.
> 
> [LLM Semantic Conventions Working Goup](https://github.com/open-telemetry/community/blob/main/projects/llm-semconv.md)
> [OTEL Semantic Conventions Sigs Index](https://github.com/open-telemetry/community/blob/main/README.md#specification-sigs)
> [Slack Channel](https://cloud-native.slack.com/archives/C06KR7ARS3X)
> 
> @drewby @pamelafox
> 
> @[gingters](https://github.com/gingters)
> gingters commented [on Apr 9](https://github.com/langfuse/langfuse/issues/199#issuecomment-2044331923)
> I would also love to see support of OpenTelemetry ingestion.
> My specific reason would be to enable observablity of [Semantic Kernel](https://github.com/microsoft/semantic-kernel) based applications in langfuse. It would be awesome if we could see the single steps of SK applications in the same way we can currently see langchain-based application traces.
> 
> @[benjamincudi](https://github.com/benjamincudi)
> benjamincudi commented [on Jul 11](https://github.com/langfuse/langfuse/issues/199#issuecomment-2222996611)
> I'd also love to see support in some kind for this - in particular, I'd love to be able to incorporate (or leverage) Langfuse tracing as part of broader application tracing, and send those traces through an OpenTelemetry Collector rather than directly to LANGFUSE_HOST
> 
> Is this something that has been tested or attempted previously with any success?
> 
> @[athewsey](https://github.com/athewsey)
> athewsey commented [on Aug 19](https://github.com/langfuse/langfuse/issues/199#issuecomment-2295557043) • 
> Also interested in this feature - my (customers') use-case would be similar to the above: Integrate Langfuse's traces with those from other microservices/apps in AWS X-Ray, for end-to-end debugging & visibility.
> 
> @[johnson-liang](https://github.com/johnson-liang)
> johnson-liang commented [on Sep 13](https://github.com/langfuse/langfuse/issues/199#issuecomment-2348823438)
> When it comes to Langfuse ingesting OpenTelemetry, I think it is more straightforward to implement an [OpenTelemetry exporter](https://opentelemetry.io/ecosystem/registry/?component=exporter) that calls Langfuse SDK for each OpenTelemetry trace.

**Comment by cubxxw** [comment]
I am wondering whether the semantic standards and compatibility of OTel have reached a level of maturity to meet practical needs. Considering the potential differences between various implementations, can this standardization progress provide sufficient stability and consistency to support a wide range of use cases?

**Reply by codefromthecrypt** [reply]
I'm a member of the community, but not yet a spec approver or lead. Here's my 2p for what it's worth:
* for observing basic chat completions and embeddings traces, metrics and logs, there's some value from system POV (think dashboards that count tokens).
* there's a cavaet in logs as many eval products do not accept otel's log format, yet.
* specs have focused so far on LLM access, while [agentic use](https://github.com/open-telemetry/semantic-conventions/pull/1657) or [evaluations](https://github.com/open-telemetry/semantic-conventions/pull/1322) are a work in progress
* things could be driven more closely to end requirements with more participation not just in issues, but also in [zoom meetings](https://github.com/open-telemetry/community/pull/2437), so please help if you can.
* the spec will call itself experimental for the foreseeable future, but that doesn't mean people won't use it.

I think LLMOps is tricky as many people think of observability just like distributed systems with certain types of metrics etc, and think the problem is solved, while folks on the MLOps side may see lack of quality metrics, or specs around training pipelines, etc and feel it isn't.

Hope I was fair and balanced in above. In summary, folks working on the spec anchor what ends up there. With more folks saying "what is practical", it is possible to have specs achieve that.

**Comment by Wh1isper** [comment]
I would like to provide an example of the necessity of supporting OTel. [Pydantic-ai](https://github.com/pydantic/pydantic-ai) is a library developed  for LLM Agent by Pydantic(which should be familiar to anyone experienced with Python development). By default, it uses [Logfire](https://github.com/pydantic/logfire) as the observability SDK. In a recent patch, the Logfire SDK has been updated to send traces to any OTel server(https://github.com/pydantic/logfire/issues/78#issuecomment-2562809216). Once Langfuse supports OTel, Pydantic-ai will be able to send traces to Langfuse without the need to hack any `completion` method.

Thanks for you work!

**Reply by marcklingen** [reply]
thanks for sharing!

**Reply by ChristianWeyer** [reply]
> thanks for sharing!

Does this change the outlook for Langfuse wrt OTel? ;-)

**Reply by marcklingen** [reply]
Not really, this is something that is very actively discussed already and this post adds another upvote that it might be the right time to finally add an initial Otel collector to Langfuse.

**Comment by marcklingen** [comment]
+1 on adding an Otel collector: Pydantic AI supports OTel instrumentation, requested here: https://github.com/orgs/langfuse/discussions/5036

cc @Steffen911

**Reply by marcklingen** [reply]
fyi, @mstrathman and @nguyenth-sieuviet as you were interested in Pydantic AI support, this will be available soon

**Comment by marcklingen** [comment]
Additional use case: [microsoft/autogen](https://github.com/microsoft/autogen)
Otel instrumentation: https://microsoft.github.io/autogen/dev//user-guide/core-user-guide/framework/telemetry.html
Thread requesting support in Langfuse by @Canada-wet: https://github.com/orgs/langfuse/discussions/5199

cc @Steffen911

**Comment by marcklingen** [comment]
Additional use case: instrumentation for Java applications
Requested here: https://github.com/orgs/langfuse/discussions/4589

**Comment by Steffen911** [comment]
Hey everyone,
On Friday, we deployed experimental OTel Support within Langfuse. You can find more details on how to send OTel Spans to Langfuse in https://langfuse.com/docs/opentelemetry/get-started. 🚀 

Please check it out and let us know if there is a need to map additional properties and if you wish to see other enhancements.

**Reply by clemra** [reply]
thank you for shipping this, steffen! very excited to hear feedback on this.

**Reply by mjrusso** [reply]
Thanks Langfuse team for implementing this!

I just tried this out in an app (that uses Magnetic, which itself uses Logfire) and it worked on the first try. The only thing that isn't ideal (for my use case at least) is that each span is assigned a separate trace, but I'd like them to all be part of the same trace.

A bit more context: my app uses the Langfuse Python SDK and the `observe` decorator. The Magnetic calls (which are OTel-instrumented) happen within observed spans. Is there any way for those calls to be considered part of the top-level `observe`d trace?

**Reply by Steffen911** [reply]
@mjrusso We use the traceId of the root span to decide whether they have the same trace or a different trace which should be the native OTel behaviour. If all your spans have the same OTel Trace-Id, they should all be in the same group. 

**Edit:** Ah, I see what you mean now. Unfortunately, the Langfuse SDK does not connect to OTel which means that they use separate TraceIds. @hassiebp may have an idea on whether we can connect the SDKs to OTel in some way, but I don't think that this would be a quick addition. Initially, we assumed that one would either use the Langfuse SDK _or_ OTel for instrumentation.

**Reply by Steffen911** [reply]
Hey @mjrusso ,
We discussed this within the team and for your purpose I'd recommend to use https://langfuse.com/docs/sdk/python/decorators#set-parent-trace-id-or-parent-span-id. With this setup you can pass the active OTel Trace Id as the TraceId into Langfuse and that way all Langfuse spans should be correctly linked together with your OTel spans.

**Reply by mjrusso** [reply]
@Steffen911 my top-level trace is created by the Langfuse SDK before any OTel tracing happens, so there's no OTel trace ID to set as `langfuse_parent_trace_id`.

What if I were to generate my own parent trace ID, and a) set that value as `langfuse_parent_trace_id`, and b) ensure that same trace ID is set as the parent for all OTel traces? (I'm assuming the latter is possible.)

**Reply by Steffen911** [reply]
@mjrusso Yes that should work. You can probably create a custom "context" and use that to create a new span with the correct traceId.

**Reply by mjrusso** [reply]
Here's what I have so far:

```python
with logfire.span("Root operation") as root_span:
    parent_trace_id = root_span.get_span_context().trace_id

    # Open Telemetry reports the trace ID as a 32-byte hexadecimal string!
    parent_trace_id_hex = format(parent_trace_id, "032x")

    compute_response(
        foo,
        bar,
        langfuse_parent_trace_id=parent_trace_id_hex,
    )
```

Where `compute_response` is defined as:

```python
    @observe  # Trace function execution with Langfuse.
    def compute_response(
        foo: Foo,
        bar: Bar,
        *,
        **kwargs,
    ):
        langfuse_context.update_current_trace(
            name="Graph Invocation",
            session_id=str(Foo.id),
        )

        ...snip...
```

(Burnt too much time on Open Telemetry reporting the ID in hex 🤦)

On the Langfuse end, all spans show up together in the same trace. They are nested in a different parent span (vs. the Langfuse-native spans), so the overall trace isn't really "in order".

**Reply by mjrusso** [reply]
Ah, let me try setting `langfuse_parent_observation_id`, too.

**Reply by mjrusso** [reply]
> Ah, let me try setting `langfuse_parent_observation_id`, too.

The OTel span doesn't know about the span created by Langfuse (via `@observe`, in this case), so the Langfuse spans and OTel spans won't share a parent, and thus won't appear in their natural order. There might be a way to give Logfire the context it needs but I'm not 100% sure; the API is confusing. (And not the end of the world; at least everything is grouped together now in the same trace! That's a huge improvement. Thanks @Steffen911 for your help!)

**Reply by Steffen911** [reply]
@mjrusso Thank you for the feedback and the investigation. We'll take a look into the Langfuse SDK - OTel connection as one of the next projects and will keep that feedback in mind. We'll keep you posted here if we make improvements in that direction.

Glad to hear that you found a setup that's already good for you 🙌

**Comment by RustamAkimov** [comment]
Thank you very much for providing this feature! 

When trying OpenLit example with self hosted version, getting this error RangeError: Invalid time value
<img width="1440" alt="image" src="https://github.com/user-attachments/assets/82ba3373-ac8e-40d1-a690-771eb955ba17" />

**Reply by RustamAkimov** [reply]
Please, skip it, i just pulled today's update and rerun docker - working now.
Leave this message, as maybe somebody will encounter similar issue.

Thank you again!

**Reply by marcklingen** [reply]
This was an intermediary issue, should be fixed in latest release, great to hear that this works for you!

**Reply by iamhatesz** [reply]
for the self-hosted version, what's the OTLP endpoint? `http://localhost:3000/api/public/otel`doesn't seem to work

**Reply by Steffen911** [reply]
@iamhatesz Which version of the self-hosted version are you running? OTel should work starting from v3.24.0.

**Reply by Wh1isper** [reply]
@iamhatesz When using `logfire`, configure like below

```bash
OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=http://localhost:3000/api/public/otel/v1/traces
OTEL_EXPORTER_OTLP_HEADERS="Authorization=Basic {AUTH_HEADER}"
```

The `AUTH_HEADER` can be generated as below:

`echo -n "pk-lf-1234:sk-lf-7890" | base64`

**Reply by iamhatesz** [reply]
Thanks. I wanted to integrate smolagents with Langfuse, but it didn't work. However, I think there's something wrong on their side.

**Reply by jannikmaierhoefer** [reply]
> Thanks. I wanted to integrate smolagents with Langfuse, but it didn't work. However, I think there's something wrong on their side.

Hi @iamhatesz, happy to help, could you you open a new GitHub post for this?

**Comment by Wh1isper** [comment]
Thank you for bringing us this feature! I’d like to know if Langfuse currently plans to support (or already supports—please forgive my ignorance) the OTEL Gen AI standard?

> https://opentelemetry.io/blog/2024/otel-generative-ai/

![image](https://github.com/user-attachments/assets/461d0413-cbe5-454c-85ff-04cd39079888)

Additionally, if I want to use Langfuse's features(with logfire) for model cost, latency, etc., the only way for now might still be manual adaptation, in `agent` layer.

```python
from pydantic_ai import Agent
import logfire

logfire.configure(
    # Setting a service name is good practice in general, but especially
    # important for Jaeger, otherwise spans will be labeled as 'unknown_service'
    service_name="trail",
    # Sending to Logfire is on by default regardless of the OTEL env vars.
    # Keep this line here if you don't want to send to both Jaeger and Logfire.
    send_to_logfire=False,
)

# Define a very simple agent including the model to use, you can also set the model when running the agent.
agent = Agent(
    "google-gla:gemini-1.5-flash",
    # Register a static system prompt using a keyword argument to the agent.
    # For more complex dynamically-generated system prompts, see the example below.
    system_prompt="Be concise, reply with one sentence.",
)

# Run the agent synchronously, conducting a conversation with the LLM.
# Here the exchange should be very short: PydanticAI will send the system prompt and the user query to the LLM,
# the model will return a text response. See below for a more complex run.
result = agent.run_sync('Where does "hello world" come from?')
print(result.data)
"""
The first known use of "hello, world" was in a 1974 textbook about the C programming language.
"""
```

**Reply by marcklingen** [reply]
Yes, we plan to extend the coverage for vendor-specific semantics over time. Please create a github issue with an example (code and link to public trace) which helps to improve the parsing of the different semantics.

**Reply by Wh1isper** [reply]
@marcklingen Thanks. 

In pydantic 0.0.23(spec PR https://github.com/pydantic/pydantic-ai/commit/35a097d0a667114d3351690cdb09f1df86aad320), we can easily the `Model` for observation. I created a simple wrapper for this: https://github.com/Wh1isper/langfuse-pydantic-ai, which solves my needs nicely.

Maybe we should add this usage to the documentation, also continue to discuss OTEL Trace with logfire support.

**Reply by Wh1isper** [reply]
Hello guys, after some investigation I found out that pydantic-ai uses a customized set of trace specs. And they are working on switching to OTEL specs: https://github.com/pydantic/pydantic-ai/issues/875#issuecomment-2646567377

I think users can use wrapper for now and switch to OTEL after that.

**Reply by marcklingen** [reply]
Great to hear this, thanks for sharing!

**Comment by Su-hub9** [comment]
Thank you! This is a really essential feature!

I am using the self-hosted version v3.24.0.
Is HTTP still not supported? When making a request to `/api/public/otel/v1/trace` using curl, the error "Failed to parse OTel Trace" occurs.
The request body is as follows.
```
{
  "resourceSpans": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "my.service"
            }
          }
        ]
      },
      "scopeSpans": [
        {
          "scope": {
            "name": "my.library",
            "version": "1.0.0",
            "attributes": [
              {
                "key": "my.scope.attribute",
                "value": {
                  "stringValue": "some scope attribute"
                }
              }
            ]
          },
          "spans": [
            {
              "traceId": "5B8EFFF798038103D269B633813FC60C",
              "spanId": "EEE19B7EC3C1B174",
              "parentSpanId": "EEE19B7EC3C1B173",
              "name": "I'm a server span",
              "startTimeUnixNano": "1544712660000000000",
              "endTimeUnixNano": "1544712661000000000",
              "kind": 2,
              "attributes": [
                {
                  "key": "my.span.attr",
                  "value": {
                    "stringValue": "some value"
                  }
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

**Reply by Steffen911** [reply]
Hey @Su-hub9 ,
For the initial support, we focused on HTTP Protobuf bodies. The `application/json` content type is not supported yet. Could you create an issue to track this?

**Reply by Su-hub9** [reply]
Thank you for the answer. I will create an issue.

Additionally, I followed the instructions in [this document](https://langfuse.com/docs/opentelemetry/get-started#export-from-opentelemetry-collector) to connect the OpenTelemetry Collector.
However, it does not run due to the following error.
```
Error: invalid configuration: exporters::otlp/langfuse: address <my-domain>/api/public/otel: missing port in address
```
```
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
 
processors:
  batch:
  memory_limiter:
    # 80% of maximum memory up to 2G
    limit_mib: 1500
    # 25% of limit up to 2G
    spike_limit_mib: 512
    check_interval: 5s
 
exporters:
  debug:
  otlp/langfuse:
    endpoint: "<my-domain>/api/public/otel"
    headers:
      Authorization: "Basic ${AUTH_STRING}"
 
service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [memory_limiter, batch]
      exporters: [debug, otlp/langfuse]
```
could there be a problem with my configuration?

**Reply by marcklingen** [reply]
Did you try adding the port? e.g. `localhost:3000/api/public/otel` (devserver)

**Reply by Su-hub9** [reply]
Sure! If you add the port, the following error occurs.
```
Error: invalid configuration: exporters::otlp/langfuse: invalid port "3000/api/public/otel"
```

**Reply by marcklingen** [reply]
How do you access the Langfuse ui? Can you share the full url?

**Reply by Su-hub9** [reply]
Langfuse UI is access at `localhost:3000`

**Reply by Steffen911** [reply]
@Su-hub9 I double checked our config and I believe we wrongly use `otlp/langfuse` instead of `otlphttp/langfuse`. As we're not supporting the gRPC protocol yet, we need to use the http interface of the exporter. I update the docs in https://github.com/langfuse/langfuse-docs/pull/1265. Could you check the changes in https://github.com/langfuse/langfuse-docs/pull/1265/files and see if they work for you?

**Reply by Su-hub9** [reply]
I tried this, and the same error as before, `"Failed to parse OTel Trace"` occurred.

**Reply by Steffen911** [reply]
@Su-hub9 Can you see more details in your server logs or try to make the same call attempt against Langfuse cloud and share the proejctId?

**Reply by Steffen911** [reply]
@Su-hub9 We'll add support for application/json type spans via https://github.com/langfuse/langfuse/pull/5450. It should go live later this evening on Langfuse cloud and will be included in our next minor release.

**Reply by Su-hub9** [reply]
Thank you for your quick feedback and assistance.

Unfortunately, the issue has not been resolved.
A parsing error occurs with the Protobuf content type. The data is in the same OTLP format as before.

The error log is as follows.
```
error  Failed to parse OTel Protobuf invalid wire type 7 at offset 1
Error: invalid wire type 7 at offset 1
    at Reader.skipType (/app/node_modules/.pnpm/protobufjs@7.4.0/node_modules/protobufjs/src/reader.js:382:19)
    at e.decode (/app/web/.next/server/chunks/5528.js:1:57098)
    at Object.fn (/app/web/.next/server/pages/api/public/otel/v1/traces.js:1:3874)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.POST (/app/web/.next/server/pages/api/public/otel/v1/traces.js:1:9806)
    at async /app/web/.next/server/chunks/802.js:1:10910
    at async /app/node_modules/.pnpm/@sentry+nextjs@8.55.0_@opentelemetry+context-async-hooks@1.30.1_@opentelemetry+api@1.9.0__@op_o6telllunn2mpmjbdvmmbwk5xm/node_modules/@sentry/nextjs/build/cjs/common/pages-router-instrumentation/wrapApiHandlerWithSentry.js:89:28
    at async K (/app/node_modules/.pnpm/next@14.2.21_@babel+core@7.24.3_@opentelemetry+api@1.9.0_@playwright+test@1.47.2_babel-plugin_yqlzwwxnhgfuesbfmfmtmhekse/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:16881)
    at async U.render (/app/node_modules/.pnpm/next@14.2.21_@babel+core@7.24.3_@opentelemetry+api@1.9.0_@playwright+test@1.47.2_babel-plugin_yqlzwwxnhgfuesbfmfmtmhekse/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:17520)
    at async NextNodeServer.runApi (/app/node_modules/.pnpm/next@14.2.21_@babel+core@7.24.3_@opentelemetry+api@1.9.0_@playwright+test@1.47.2_babel-plugin_yqlzwwxnhgfuesbfmfmtmhekse/node_modules/next/dist/server/next-server.js:598:9)
```

**Reply by Steffen911** [reply]
@Su-hub9 Can you log the trace that you emit? That would help us reconstruct the error as the error message isn't very helpful here.

**Reply by Su-hub9** [reply]
I'm not sure what kind of logs you're asking for. The request is being sent with the same data as the comment I initially made.
I'll include the OpenTelemetry Collector's debug log.

```
info	Traces	{"kind": "exporter", "data_type": "traces", "name": "debug", "resource spans": 2, "spans": 2}
debug	otlphttpexporter@v0.118.0/otlp.go:177	Preparing to make HTTP request	{"kind": "exporter", "data_type": "traces", "name": "otlphttp/langfuse", "url": "http://localhost:3000/api/public/otel/v1/traces"}
info	ResourceSpans #0
Resource SchemaURL: 
Resource attributes:
     -> telemetry.sdk.language: Str(python)
     -> telemetry.sdk.name: Str(opentelemetry)
     -> telemetry.sdk.version: Str(1.30.0)
     -> service.name: Str(my-python-app)
ScopeSpans #0
ScopeSpans SchemaURL: 
InstrumentationScope __main__ 
Span #0
    Trace ID       : ef278c94a45d4527ba17c0cd5750bbc2
    Parent ID      : e260073e994c600c
    ID             : 992f9a7d4315812d
    Name           : hello_span
    Kind           : Internal
    Start time     : 2025-02-22 00:53:01.546362 +0000 UTC
    End time       : 2025-02-22 00:53:01.546382 +0000 UTC
    Status code    : Unset
    Status message : 
ResourceSpans #1
Resource SchemaURL: 
Resource attributes:
     -> telemetry.sdk.language: Str(python)
     -> telemetry.sdk.name: Str(opentelemetry)
     -> telemetry.sdk.version: Str(1.30.0)
     -> service.name: Str(my-python-app)
ScopeSpans #0
ScopeSpans SchemaURL: 
InstrumentationScope opentelemetry.instrumentation.flask 0.51b0
Span #0
    Trace ID       : ef278c94a45d4527ba17c0cd5750bbc2
    Parent ID      : 
    ID             : e260073e994c600c
    Name           : POST /trace
    Kind           : Server
    Start time     : 2025-02-22 00:53:01.544921 +0000 UTC
    End time       : 2025-02-22 00:53:02.55939 +0000 UTC
    Status code    : Unset
    Status message : 
Attributes:
     -> http.method: Str(POST)
     -> http.server_name: Str(0.0.0.0)
     -> http.scheme: Str(http)
	{"kind": "exporter", "data_type": "traces", "name": "debug"}
error	internal/queue_sender.go:105	Exporting failed. Dropping data.	{"kind": "exporter", "data_type": "traces", "name": "otlphttp/langfuse", "error": "not retryable error: Permanent error: rpc error: code = InvalidArgument desc = error exporting items, request to http://localhost:3000/api/public/otel/v1/traces responded with HTTP Status Code 400", "dropped_items": 2}
```

**Reply by Steffen911** [reply]
@Su-hub9 Can you share your full OTel collector version and configuration? Perhaps this helps me to reproduce the case.

**Reply by Su-hub9** [reply]
The OpenTelemetry Collector version is 0.118.0. I'm sharing the config setup.
```
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
processors:
  batch:
exporters:
  debug:
    verbosity: detailed
  otlphttp/langfuse:
    endpoint: "http://localhost:3000/api/public/otel"
    headers:
      Authorization: "Basic <API-KEY>"
service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [debug, otlphttp/langfuse]
  telemetry:
    logs:
      level: "debug"
```

**Reply by Steffen911** [reply]
Hey @Su-hub9 ,
Thank you for providing the configuration. I was able to reproduce it using your settings. It turns out that the otel collector uses gzip compression by default and our endpoint did not handle that correctly. This will be fixed with https://github.com/langfuse/langfuse/pull/5710 which should go live on our cloud version later today. 

In the meantime, you can work around this by setting `compression: none` as documented in https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/otlphttpexporter.

**Reply by liefra** [reply]
Hi, 

I have been reading your [structured output](https://langfuse.com/docs/integrations/openai/python/structured-outputs#see-your-trace-in-langfuse) information and watched the recent tool calling and structured outputs in the Langfuse playground YouTube video, and I like the feature but cannot find a way to **submit the `response_format` structured outputs using OpenTelemetry** on the [OpenTelemetry property mapping](https://langfuse.com/docs/opentelemetry/get-started#property-mapping) page.

Does anyone know whether this is currently working? 

If so, a code snippet showing the OpenTelemetry attribute keys would be much appreciated.

**Reply by marcklingen** [reply]
They are included within the input/output in langfuse tracing.

**Reply by liefra** [reply]
Hi @marcklingen, thanks for your reply. It’s working now after setting the attribute on a span to the following key: 

`langfuse.metadata.response_format`

NOTE: Note, I am using the Java OpenTelemetry SDK. 
```java
.setAttribute("langfuse.metadata.response_format", jsonString)
```

**Comment by xiaoping378** [comment]
https://opentelemetry.io/blog/2024/otel-generative-ai/
langfuse don't support events of GEN-AI ？

**Reply by Steffen911** [reply]
@xiaoping378 As of now, we recommend to still use span attributes instead of events as this makes the correlation to other properties significantly easier.
Could you share more details about your instrumentation and how your spans are generated?

**Reply by xiaoping378** [reply]
used zero-code [instrumentation](https://opentelemetry.io/docs/zero-code/python/) , now i need choose a sdk instrumentation

**Reply by ghost** [reply]
any update on this? i'm sending `gen_ai.output.messages` and not seeing it be respected. How do i send you the response with otel?

**Reply by Steffen911** [reply]
@krrishdholakia Could you share the events you're sending or a small reproduction? Currently standalone events are not supported, but if you send them as Span events they should be included.
